File: classroom_app/schemas/api_common.py
from __future__ import annotations

from enum import Enum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class ApiErrorCode(str, Enum):
    BAD_REQUEST = "bad_request"
    LOGIN_REQUIRED = "login_required"
    PERMISSION_DENIED = "permission_denied"
    NOT_FOUND = "not_found"
    VALIDATION_ERROR = "validation_error"
    CONFLICT = "conflict"
    RATE_LIMITED = "rate_limited"
    PAYLOAD_TOO_LARGE = "payload_too_large"
    UNSUPPORTED_MEDIA_TYPE = "unsupported_media_type"
    SERVICE_UNAVAILABLE = "service_unavailable"
    UPSTREAM_ERROR = "upstream_error"
    INTERNAL_ERROR = "internal_error"
    UNKNOWN_ERROR = "unknown_error"


API_ERROR_CODE_BY_STATUS: dict[int, ApiErrorCode] = {
    400: ApiErrorCode.BAD_REQUEST,
    401: ApiErrorCode.LOGIN_REQUIRED,
    403: ApiErrorCode.PERMISSION_DENIED,
    404: ApiErrorCode.NOT_FOUND,
    409: ApiErrorCode.CONFLICT,
    413: ApiErrorCode.PAYLOAD_TOO_LARGE,
    415: ApiErrorCode.UNSUPPORTED_MEDIA_TYPE,
    422: ApiErrorCode.VALIDATION_ERROR,
    429: ApiErrorCode.RATE_LIMITED,
    500: ApiErrorCode.INTERNAL_ERROR,
    502: ApiErrorCode.UPSTREAM_ERROR,
    503: ApiErrorCode.SERVICE_UNAVAILABLE,
    504: ApiErrorCode.UPSTREAM_ERROR,
}


class ApiFlexibleRecord(BaseModel):
    model_config = ConfigDict(extra="allow")


class ApiStatusResponse(BaseModel):
    model_config = ConfigDict(extra="allow")

    status: str


class ApiSuccessResponse(ApiStatusResponse):
    status: str = "success"


class ApiErrorBody(ApiFlexibleRecord):
    code: ApiErrorCode
    message: str = Field(min_length=1)
    details: dict[str, Any] | None = None
    request_id: str | None = None


class ApiErrorResponse(ApiFlexibleRecord):
    detail: str | dict[str, Any] | list[Any] | None = None
    code: ApiErrorCode
    error: ApiErrorBody


def api_error_code_for_status(status_code: int) -> ApiErrorCode:
    return API_ERROR_CODE_BY_STATUS.get(int(status_code), ApiErrorCode.UNKNOWN_ERROR)


def error_message_from_detail(detail: Any, fallback: str) -> str:
    if isinstance(detail, str) and detail.strip():
        return detail.strip()
    if isinstance(detail, dict):
        for key in ("message", "error", "reason", "detail"):
            value = detail.get(key)
            if isinstance(value, str) and value.strip():
                return value.strip()
    return fallback


def build_error_payload(
    *,
    detail: str | dict[str, Any] | list[Any] | None,
    message: str,
    code: ApiErrorCode,
    details: dict[str, Any] | None = None,
    request_id: str | None = None,
    legacy_fields: dict[str, Any] | None = None,
) -> dict[str, Any]:
    error: dict[str, Any] = {
        "code": code.value,
        "message": message,
    }
    if details:
        error["details"] = details
    if request_id:
        error["request_id"] = request_id

    payload: dict[str, Any] = {
        "detail": detail,
        "code": code.value,
        "error": error,
    }
    if legacy_fields:
        payload.update(legacy_fields)
    return payload

File: frontend/src/contracts/api-common.ts
import { z } from 'zod';

export const ApiErrorCodeSchema = z.enum([
  'bad_request',
  'login_required',
  'permission_denied',
  'not_found',
  'validation_error',
  'conflict',
  'rate_limited',
  'payload_too_large',
  'unsupported_media_type',
  'service_unavailable',
  'upstream_error',
  'internal_error',
  'unknown_error',
]);

export type ApiErrorCode = z.infer<typeof ApiErrorCodeSchema>;

export const FlexibleRecordSchema = z.object({}).catchall(z.unknown());
export type FlexibleRecord = z.infer<typeof FlexibleRecordSchema>;

export const ApiErrorBodySchema = z.object({
  code: ApiErrorCodeSchema,
  message: z.string().min(1),
  details: z.record(z.string(), z.unknown()).optional(),
  request_id: z.string().nullable().optional(),
}).passthrough();

export const ApiErrorPayloadSchema = z.object({
  detail: z.unknown().optional(),
  code: ApiErrorCodeSchema.optional(),
  error: ApiErrorBodySchema.optional(),
}).passthrough();

export type ApiErrorPayload = z.infer<typeof ApiErrorPayloadSchema>;

export function messageFromUnknownDetail(detail: unknown): string | null {
  if (typeof detail === 'string' && detail.trim()) {
    return detail.trim();
  }
  if (detail && typeof detail === 'object') {
    const record = detail as Record<string, unknown>;
    for (const key of ['message', 'error', 'reason', 'detail']) {
      const value = record[key];
      if (typeof value === 'string' && value.trim()) {
        return value.trim();
      }
    }
  }
  return null;
}

export function parseApiErrorPayload(payload: unknown): {
  code?: ApiErrorCode;
  message?: string;
  details?: Record<string, unknown>;
  requestId?: string | null;
} {
  const parsed = ApiErrorPayloadSchema.safeParse(payload);
  if (!parsed.success) {
    return {
      message: messageFromUnknownDetail(payload) ?? undefined,
    };
  }

  const error = parsed.data.error;
  return {
    code: error?.code ?? parsed.data.code,
    message: error?.message ?? messageFromUnknownDetail(parsed.data.detail) ?? undefined,
    details: error?.details,
    requestId: error?.request_id ?? undefined,
  };
}

export const ApiSuccessStatusSchema = z.literal('success');
